File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/Activities/LoginActivity.java
package kore.botssdk.audiocodes.webrtcclient.Activities;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.audiocodes.mv.webrtcsdk.log.Log;
import com.audiocodes.mv.webrtcsdk.sip.enums.Transport;

import kore.botssdk.R;
import kore.botssdk.application.BotApplication;
import kore.botssdk.audiocodes.oauth.OAuthManager;
import kore.botssdk.audiocodes.webrtcclient.General.ACManager;
import kore.botssdk.audiocodes.webrtcclient.General.AppUtils;
import kore.botssdk.audiocodes.webrtcclient.General.Prefs;
import kore.botssdk.audiocodes.webrtcclient.Structure.SipAccount;


public class LoginActivity extends BaseAppCompatActivity {

    public static final String OAUTH_FAILED = "OAUTH_FAILED";

    private boolean oAuthEnable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.login_activity);

        initGui();
    }

    private void initGui() {

        TextView outhFailedTV = findViewById(R.id.login_textview_outh_failed);
        if (getIntent().getBooleanExtra(OAUTH_FAILED, false)) {
            outhFailedTV.setVisibility(View.VISIBLE);
        } else {
            outhFailedTV.setVisibility(View.GONE);
        }

        SipAccount sipAccount = new SipAccount();


        EditText userNameEditText = (EditText) findViewById(R.id.login_editText_username);
        //userNameEditText.setText(sipAccount.getUsername());
        EditText passwordEditText = (EditText) findViewById(R.id.login_editText_password);
        EditText displayNameEditText = (EditText) findViewById(R.id.login_editText_displayname);
        EditText domainEditText = (EditText) findViewById(R.id.login_editText_domain);
        EditText sipAddressEditText = (EditText) findViewById(R.id.login_editText_sipaddress);
        EditText portEditText = (EditText) findViewById(R.id.login_editText_port);
        Spinner transportSpinner = (Spinner) findViewById(R.id.login_editText_transport);


        transportSpinner.setAdapter(new ArrayAdapter<Transport>(this, android.R.layout.simple_spinner_item, Transport.values()));
//        EditText turnServerEditText = (EditText)findViewById(R.id.login_editText_turnserver);
//        EditText stunserverEditText = (EditText)findViewById(R.id.login_editText_stunserver);

        Button loginButton = (Button) findViewById(R.id.login_button_login);
        CheckBox autologin = (CheckBox) findViewById(R.id.autologin_checkBox);
        autologin.setChecked(true);

        CheckBox disconnectCall = (CheckBox) findViewById(R.id.call_disconnect_checkBox);
        disconnectCall.setChecked(true);

        CheckBox usePush = (CheckBox) findViewById(R.id.push_checkBox);
        usePush.setChecked(Prefs.usePush());

        usePush.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                Prefs.setUsePush(isChecked);
            }
        });

        CheckBox oauthCB = (CheckBox) findViewById(R.id.oauth_checkBox);
        oauthCB.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                LinearLayout layout = (LinearLayout) findViewById(R.id.oauth_layout);
                oAuthEnable = isChecked;
                if (isChecked) {
                    layout.setVisibility(View.VISIBLE);
                } else {
                    layout.setVisibility(View.GONE);
                }
            }
        });

        CheckBox useVideoHardware = (CheckBox) findViewById(R.id.video_hardware_checkbox);
        useVideoHardware.setChecked(Prefs.isVideoHardware());

        useVideoHardware.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                Prefs.setVideoHardware(isChecked);
            }
        });

        EditText oauthUrlEditText = (EditText) findViewById(R.id.login_editText_oauth_url);
        EditText oauthRealmEditText = (EditText) findViewById(R.id.login_editText_oauth_realm);
        EditText oauthClientIdEditText = (EditText) findViewById(R.id.login_editText_oauth_clientid);

        if (OAuthManager.getInstance().isEnabled()) {
            Log.d("BBB", "getting in here");
            oauthCB.setVisibility(View.GONE);
            oAuthEnable = true;
            LinearLayout layout = (LinearLayout) findViewById(R.id.oauth_layout);
            layout.setVisibility(View.VISIBLE);
        }

        boolean debugMode = AppUtils.getStringBoolean(R.string.enable_debug_mode);
        if (debugMode) {
            userNameEditText.setText(sipAccount.getUsername());
            passwordEditText.setText(sipAccount.getPassword());
            displayNameEditText.setText(sipAccount.getDisplayName());
            domainEditText.setText(sipAccount.getDomain());
            sipAddressEditText.setText(sipAccount.getProxy());

            for (int i = 0; i < transportSpinner.getAdapter().getCount(); i++) {
                if (transportSpinner.getItemAtPosition(i).toString().toLowerCase().equals(sipAccount.getTransport().toString().toLowerCase())) {
                    transportSpinner.setSelection(i);
                }
            }

            //userNameEditText.setText(sipAccount.getUsername());
            //userNameEditText.setText(sipAccount.getUsername());
            portEditText.setText(String.valueOf(sipAccount.getPort()));

            //advancedInfoLayout.setVisibility( View.VISIBLE );
            oauthCB.setChecked(AppUtils.getStringBoolean(R.string.oauth_enable));
            oauthUrlEditText.setText(getString(R.string.oauth_url));
            oauthRealmEditText.setText(getString(R.string.oauth_realm));
            oauthClientIdEditText.setText(getString(R.string.oauth_client_id));
        }


        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                if (validateField(userNameEditText) && validateField(passwordEditText)) {

                    String userName = userNameEditText.getText().toString().trim();
                    String password = passwordEditText.getText().toString().trim();
                    sipAccount.setUsername(userName);
                    if (!oAuthEnable) {
                        sipAccount.setPassword(password);
                    } else {
                        sipAccount.setPassword("no password");
                    }
                    //if(validateField(displayNameEditText)&&validateField(domainEditText)&&validateField(sipAddressEditText)&&validateField(turnServerEditText)&&validateField(stunserverEditText))
                    if (validateField(displayNameEditText) && validateField(domainEditText) && validateField(sipAddressEditText) && validateField(portEditText)) {
                        sipAccount.setDisplayName(displayNameEditText.getText().toString().trim());
                        sipAccount.setDomain(domainEditText.getText().toString().trim());
                        sipAccount.setProxy(sipAddressEditText.getText().toString().trim());
                        sipAccount.setTransport(AppUtils.getTransport(transportSpinner.getSelectedItem().toString().trim()));

                        int portNumber;
                        try {
                            portNumber = Integer.valueOf(portEditText.getText().toString().trim());
                        } catch (Exception e) {
                            //use default port number;
                            portNumber = Integer.valueOf(getString(R.string.sip_account_port_default));
                        }
                        sipAccount.setPort(portNumber);
                        //sipAccount.setPassword(turnServerEditText.getText().toString().trim());
                        //sipAccount.setUsername(stunserverEditText.getText().toString().trim());
                    }

                    Prefs.setSipAccount(sipAccount);
                    if (oAuthEnable) {
                        OAuthManager.getInstance().setURL(oauthUrlEditText.getText().toString());
                        OAuthManager.getInstance().setRealm(oauthRealmEditText.getText().toString());
                        OAuthManager.getInstance().setClientId(oauthClientIdEditText.getText().toString());
                        OAuthManager.getInstance().authorize(userName, password, new OAuthManager.LoginCallback() {
                            @Override
                            public void onAuthorize(boolean success) {
                                if (success) {
                                    openNextScreen(autologin.isChecked(), disconnectCall.isChecked());
                                } else {
                                    Toast.makeText(BotApplication.getGlobalContext(), getString(R.string.oauth_error), Toast.LENGTH_SHORT).show();
                                }
                            }
                        });
                    } else {
                        openNextScreen(autologin.isChecked(), disconnectCall.isChecked());
                    }
                }
                Prefs.setSipAccount(sipAccount);
                if (!oAuthEnable) {
                    openNextScreen(autologin.isChecked(), disconnectCall.isChecked());
                }
            }
        });

        boolean predefinedConfigMode = AppUtils.getStringBoolean(R.string.predefined_config_mode);
        if (predefinedConfigMode) {
            userNameEditText.setEnabled(false);
            passwordEditText.setEnabled(false);
            displayNameEditText.setEnabled(false);
            domainEditText.setEnabled(false);
            sipAddressEditText.setEnabled(false);
            portEditText.setEnabled(false);
            transportSpinner.setEnabled(false);

            //use click to call
            autologin.setChecked(false);
            autologin.setEnabled(false);

            loginButton.setText(R.string.login_button_start_text);
        }
    }

    private boolean validateField(EditText editText) {
        if (editText != null && editText.getText() != null && !editText.getText().toString().trim().equals("")) {
            return true;
        }
        return false;
    }

    private void openNextScreen(boolean autologin, boolean disconnectCall) {
        Prefs.setFirstLogin(false);
        //start login and open app main screen
        ACManager.getInstance().startLogin(autologin, disconnectCall);
        startNextActivity(MainActivity.class);
    }

}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/activities/LoginActivity.kt
package com.kore.ui.audiocodes.webrtcclient.activities

import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Toast
import androidx.core.view.isVisible
import androidx.databinding.DataBindingUtil
import com.audiocodes.mv.webrtcsdk.sip.enums.Transport
import com.kore.common.utils.LogUtils
import com.kore.ui.R
import com.kore.ui.audiocodes.oauth.OAuthManager
import com.kore.ui.audiocodes.webrtcclient.general.ACManager
import com.kore.ui.audiocodes.webrtcclient.general.AppUtils
import com.kore.ui.audiocodes.webrtcclient.general.Prefs
import com.kore.ui.audiocodes.webrtcclient.structure.SipAccount
import com.kore.ui.databinding.LoginActivityBinding
import java.util.Locale

class LoginActivity : BaseAppCompatActivity() {
    private lateinit var binding: LoginActivityBinding
    private var oAuthEnable = false
    private val acManager: ACManager = ACManager.getInstance()
    private val oAuthManager: OAuthManager = OAuthManager.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.inflate(layoutInflater, R.layout.login_activity, null, false)
        setContentView(binding.root)
        initGui()
    }

    private fun initGui() {
        binding.loginTextviewOauthFailed.isVisible = intent.getBooleanExtra(OAUTH_FAILED, false)
        val sipAccount = SipAccount(this)
        binding.loginEditTextTransport.adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, Transport.values())
        binding.autologinCheckBox.isChecked = true
        binding.callDisconnectCheckBox.isChecked = true
        binding.pushCheckBox.isChecked = Prefs.usePush(this)
        binding.pushCheckBox.setOnCheckedChangeListener { _, isChecked -> Prefs.setUsePush(this, isChecked) }
        binding.oauthCheckBox.setOnCheckedChangeListener { _, isChecked ->
            oAuthEnable = isChecked
            binding.oauthLayout.isVisible = oAuthEnable
        }
        binding.videoHardwareCheckbox.isChecked = Prefs.getIsVideoHardware(this)
        binding.videoHardwareCheckbox.setOnCheckedChangeListener { _, isChecked -> Prefs.setIsVideoHardware(this, isChecked) }
        if (oAuthManager.isEnabled) {
            LogUtils.d("BBB", "getting in here")
            binding.oauthCheckBox.visibility = View.GONE
            oAuthEnable = true
            binding.oauthLayout.isVisible = true
        }
        val debugMode = AppUtils.getStringBoolean(this, R.string.enable_debug_mode)
        if (debugMode) {
            binding.loginEditTextUsername.setText(sipAccount.username)
            binding.loginEditTextPassword.setText(sipAccount.password)
            binding.loginEditTextDisplayname.setText(sipAccount.displayName)
            binding.loginEditTextDomain.setText(sipAccount.domain)
            binding.loginEditTextSipaddress.setText(sipAccount.proxy)
            for (i in 0 until binding.loginEditTextTransport.adapter.count) {
                if (binding.loginEditTextTransport.getItemAtPosition(i).toString()
                        .lowercase(Locale.getDefault()) == sipAccount.transport.toString().lowercase(Locale.getDefault())
                ) {
                    binding.loginEditTextTransport.setSelection(i)
                }
            }

            binding.loginEditTextPort.setText(sipAccount.port.toString())
            binding.oauthCheckBox.isChecked = AppUtils.getStringBoolean(this, R.string.oauth_enable)
            binding.loginEditTextOauthUrl.setText(getString(R.string.oauth_url))
            binding.loginEditTextOauthRealm.setText(getString(R.string.oauth_realm))
            binding.loginEditTextOauthClientid.setText(getString(R.string.oauth_client_id))
        }
        binding.loginButtonLogin.setOnClickListener {
            if (validateField(binding.loginEditTextUsername) && validateField(binding.loginEditTextPassword)) {
                val userName = binding.loginEditTextUsername.text.toString().trim { it <= ' ' }
                val password = binding.loginEditTextPassword.text.toString().trim { it <= ' ' }
                sipAccount.username = userName
                sipAccount.password = if (!oAuthEnable) password else "no password"
                if (validateField(binding.loginEditTextDisplayname) && validateField(binding.loginEditTextDomain) &&
                    validateField(binding.loginEditTextSipaddress) && validateField(binding.loginEditTextPort)
                ) {
                    sipAccount.displayName = binding.loginEditTextDisplayname.text.toString().trim { it <= ' ' }
                    sipAccount.domain = binding.loginEditTextDomain.text.toString().trim { it <= ' ' }
                    sipAccount.proxy = binding.loginEditTextSipaddress.text.toString().trim { it <= ' ' }
                    sipAccount.transport = AppUtils.getTransport(this, binding.loginEditTextTransport.selectedItem.toString().trim { it <= ' ' })
                    val portNumber: Int = try {
                        Integer.valueOf(binding.loginEditTextPort.text.toString().trim { it <= ' ' })
                    } catch (e: Exception) {
                        //use default port number;
                        Integer.valueOf(getString(R.string.sip_account_port_default))
                    }
                    sipAccount.port = portNumber
                }
                Prefs.setSipAccount(this, sipAccount)
                if (oAuthEnable) {
                    oAuthManager.setURL(binding.loginEditTextOauthUrl.text.toString())
                    oAuthManager.setRealm(binding.loginEditTextOauthRealm.text.toString())
                    oAuthManager.setClientId(binding.loginEditTextOauthClientid.text.toString())
                    oAuthManager.authorize(this, userName, password, object : OAuthManager.LoginCallback {
                        override fun onAuthorize(success: Boolean) {
                            if (success) {
                                openNextScreen(binding.autologinCheckBox.isChecked, binding.callDisconnectCheckBox.isChecked)
                            } else {
                                Toast.makeText(this@LoginActivity.applicationContext, getString(R.string.oauth_error), Toast.LENGTH_SHORT).show()
                            }
                        }
                    })
                } else {
                    openNextScreen(binding.autologinCheckBox.isChecked, binding.callDisconnectCheckBox.isChecked)
                }
            }
            Prefs.setSipAccount(this, sipAccount)
            if (!oAuthEnable) {
                openNextScreen(binding.autologinCheckBox.isChecked, binding.callDisconnectCheckBox.isChecked)
            }
        }
        val predefinedConfigMode = AppUtils.getStringBoolean(this, R.string.predefined_config_mode)
        if (predefinedConfigMode) {
            binding.loginEditTextOauthUrl.isEnabled = false
            binding.loginEditTextPassword.isEnabled = false
            binding.loginEditTextDisplayname.isEnabled = false
            binding.loginEditTextDomain.isEnabled = false
            binding.loginEditTextSipaddress.isEnabled = false
            binding.loginEditTextPort.isEnabled = false
            binding.loginEditTextTransport.isEnabled = false

            //use click to call
            binding.autologinCheckBox.isChecked = false
            binding.autologinCheckBox.isEnabled = false
            binding.loginButtonLogin.setText(R.string.login_button_start_text)
        }
    }

    private fun validateField(editText: EditText?): Boolean {
        return editText != null && editText.text != null && editText.text.toString().trim { it <= ' ' } != ""
    }

    internal fun openNextScreen(autologin: Boolean, disconnectCall: Boolean) {
        Prefs.setIsFirstLogin(this, false)
        //start login and open app main screen
        acManager.startLogin(this, autologin, disconnectCall)
        startNextActivity(MainActivity::class.java)
    }

    companion object {
        const val OAUTH_FAILED = "OAUTH_FAILED"
    }
}